lib/openhab/dsl/time/month_day.rb

```diff
@@ -0,0 +1,140 @@
+        # @return [Boolean] Returns true if supplied range can be converted to a month day range
+        def self.range?(range)
+          return false unless range.is_a? Range
+
+          MonthDay.day_of_month?(range.begin) && MonthDay.day_of_month?(range.end)
+        end
+
+        # Represents a range element for a MonthDay object
+        # The LocalDate (MonthDay + Year) is dynamically calculated to allow for
+        # being used as a guard during rule evaluation
+        class MonthDayRangeElement
+          include Comparable
+          include OpenHAB::Log
+          java_import java.time.LocalDate
+          java_import java.time.Year
+
+          # Create a new MonthDayRange element
+          # @param [MonthDay] MonthDay element
+          # @param [Lambda] year lambda to calculate year to convert MonthDay to LocalDate
+          #
+          def initialize(month_day:, year:)
+            @month_day = month_day
+            @year = year
+          end
+
+          # Convert into a LocalDate using year lambda supplied in initializer
+          def to_local_date
+            @year.call.at_month_day(@month_day)
+          end
+
+          # Returns the MonthDay advanced by 1 day
+          # Required by Range class
+          def succ
+            next_date = to_local_date.plus_days(1)
+            # Handle rollover to next year
+            year = -> { Year.from(next_date) }
+            MonthDayRangeElement.new(month_day: MonthDay.from(next_date), year: year)
+          end
+
+          # Compare MonthDayRangeElement to other objects as required by Range class
+          # rubocop:disable Metrics/AbcSize
+          # Case statement needs to work against multiple types
+          def <=>(other)
+            case other
+            when LocalDate then to_local_date.compare_to(other)
+            when Date then self.<=>(LocalDate.of(other.year, other.month, other.day))
+            when MonthDay then self.<=>(MonthDayRange.current_year.call.at_month_day(other))
+            else
+              return self.<=>(other.to_local_date) if other.respond_to? :to_local_date
+              return self.<=>(other.to_date) if other.respond_to? :to_date
+
+              raise "Unable to convert #{other.class} to compare to MonthDay"
+            end
+          end
+          # rubocop:enable Metrics/AbcSize
+        end
+      end
+
+      java_import java.time.MonthDay
+      # Extend MonthDay java object with some helper methods
+      class MonthDay
+        include OpenHAB::Log
+        # Parse MonthDay string as defined with by Monthday class without leading double dash "--"
+        def self.parse(string)
+          ##          string = "--#{string}" unless string.to_s.start_with? '--'
+          java_send :parse, [java.lang.CharSequence], "--#{string}"
+        end
+
+        # Can the supplied object be parsed into a MonthDay
+        def self.day_of_month?(obj)
+          /^-*[01][0-9]-[0-3]\d$/.match? obj.to_s
```

Comment:
Why not use parse and catch the exception? This way if we ever wanted to change parse to support '6 Jan' it would still work?

Reply:
If we want to update with additional formats we would need to update this check.

Java exceptions are fairly expensive and I try to follow the effective Java guidelines.

http://www.thefinestartist.com/effective-java/57